### DISCUSSION TITLE
Cannot find module in child process when running next build

### DISCUSSION THREAD

**Discussion by Capricair** [vercel/next.js#41346] [Help]
Hi, I wrote a deploy script to run some commands with the child process of nodejs, like git checkout, npm install, npm run build, and I created an npm script named deploy, run `npm run deploy` it works well, I want run a service on server to receive a request to run `cd /to/website/directory && npm run deploy`, but it throw cannot find module error when running `next build`

```
building...
(node:1086726) UnhandledPromiseRejectionWarning: Error: Command failed: yarn run build
error - Failed to load next.config.js, see more info here https://nextjs.org/docs/messages/next-config-error

> Build error occurred
Error: Cannot find module '@next/bundle-analyzer'
Require stack:
- /path/to/website/next.config.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:889:15)
    at Function.mod._resolveFilename (/path/to/website/node_modules/next/dist/build/webpack/require-hook.js:183:28)
    at Function.Module._load (internal/modules/cjs/loader.js:745:27)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.<anonymous> (/path/to/website/next.config.js:1:28)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/path/to/website/next.config.js' ]
}
```
this module is already installed

I change to run shell file deploy.sh also doesn't work
```
cd /path/to/website/
npm run deploy
```

**Comment by icyJoseph** [comment]
Are you absolutely sure the module is there? If it is a dev dependency, and you do `npm ci` then it won't be present. I'd assume you don't have every other so called, dev dependency, installed on your runtime.

What you could try to do is:

```js
const withBundleAnalyzer =
    process.env.ANALYZE_BUNDLE === 'true' ? require('@next/bundle-analyzer')({enabled: true}) : (config) => config;
```

This way, locally, you set analyze via .env files or via `ANALYZE_BUNDLE npm run build` for example, but when deployed, and trying to re-run, you can skip the bundle analyzer.

That's of course assuming you don't want/need to analyze the bundle through this script. This happened to us a while back when doing standalone mode, where we required the bundle-analyzer, and controlled the enabled flag through an env var, but the standalone mode didn't copy the analyzer dependency, so the build crashed. 

```
issue-tester on  main [✘!?] via  v16.14.0 at  from   took 6s
➜ rm -rf node_modules/ .next/

issue-tester on  main [✘!?] via  v16.14.0 at  from 
➜ yarn install --production
yarn install v1.22.19
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
✨  Done in 1.22s.

issue-tester on  main [✘!?] via  v16.14.0 at  from 
➜ yarn build
yarn run v1.22.19
$ next build
info  - Loaded env from /path/to/my/package
error - Failed to load next.config.js, see more info here https://nextjs.org/docs/messages/next-config-error

> Build error occurred
Error: Cannot find module '@next/bundle-analyzer'
```

But with the above suggestion it'll build, unless ANALYZE is set to "true".

**Reply by Capricair** [reply]
I absolutely sure the module is installed, here is my deploy script:

```
const { exec } = require("child_process")

async function runCommand(command) {
  return await new Promise((resolve, reject) => {
    exec(command, (error, stdout) => {
      if (error) {
        reject(error)
      } else {
        console.log(stdout)
        resolve()
      }
    })
  })
}

;(async () => {
  const branch = process.env.BRANCH || "master"
  const commit = process.env.COMMIT

  console.log(`checkout git branch ${branch}...`)
  await runCommand("git remote prune origin")
  await runCommand(`git checkout ${branch}`)
  if (commit) {
    console.log(`reset to commit ${commit}`)
    await runCommand(`git reset --hard ${commit}`)
  } else {
    await runCommand(`git reset --hard`)
  }
  await runCommand("git pull")

  console.log("install packages...")
  await runCommand("yarn install")

  console.log("building...")
  await runCommand(`yarn run build`)
  console.log(`build complete!`)

  console.log("pm2 restarting...")
  await runCommand("pm2 restart ecosystem.config.js")
  console.log("deploy complete!")
})()
```

I created an api use child process run `yarn run deploy`, the strange thing is that only the next build command went wrong

I have tried removing the code related to @next/bundle-analyzer in next.config.js, but still can't find other installed modules, it should have nothing to do with @next/bundle-analyzer

**Reply by icyJoseph** [reply]
What node version have you got installed, and from which directory are you calling the script?

**Reply by Capricair** [reply]
node version: 16.16.0
The script was called from another project, run `cd /path/to/website && yarn run deploy` use the exec of the child process

**Reply by icyJoseph** [reply]
A couple of things, could you  update `npm/yarn` to latest, and run `ls node_modules` after install?

**Reply by Capricair** [reply]
OK, I'll try it, thanks!
